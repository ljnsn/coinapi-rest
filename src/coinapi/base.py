"""Base class for operation collections."""

import enum
from typing import TypeVar

import httpx
import msgspec
from httpx import codes

from coinapi import utils
from coinapi._hooks import HookContext
from coinapi.config import CoinAPIConfig
from coinapi.models import errors
from coinapi.models.operations.base import CoinAPIRequest, CoinAPIResponse


class AcceptEnum(str, enum.Enum):
    """Valid accept header values."""

    APPLICATION_JSON = "application/json"
    TEXT_JSON = "text/json"
    TEXT_PLAIN = "text/plain"
    APPLICATION_X_MSGPACK = "application/x-msgpack"
    ANY = "*/*"


RequestT = TypeVar("RequestT", bound=CoinAPIRequest)
ResponseT = TypeVar("ResponseT", bound=CoinAPIResponse)


class Base:
    """Base class for operation collections."""

    def __init__(self, sdk_config: CoinAPIConfig) -> None:
        self.sdk_configuration = sdk_config

    def _make_request(  # noqa: PLR0912, C901
        self,
        operation_id: str,
        request: RequestT,
        response_cls: type[ResponseT],
        accept_header_override: AcceptEnum | None = None,
    ) -> ResponseT:
        """Send a request."""
        hook_ctx = HookContext(
            operation_id=operation_id,
            oauth2_scopes=[],
            security_source=self.sdk_configuration.security,
        )
        base_url = utils.template_url(*self.sdk_configuration.get_server_details())
        url = utils.generate_url(type(request), base_url, request.endpoint, request)  # type: ignore[arg-type]
        headers = {}
        data, form = None, None
        if request.method in {"POST", "PUT", "PATCH"}:
            req_content_type, data, form = utils.serialize_request_body(
                request,
                "body",
            )
            if req_content_type is not None and req_content_type not in (
                "multipart/form-data",
                "multipart/mixed",
            ):
                headers["content-type"] = req_content_type
        query_params = utils.get_query_params(type(request), request) or None  # type: ignore[arg-type]
        if accept_header_override is not None:
            headers["Accept"] = accept_header_override.value
        else:
            headers["Accept"] = (
                "application/json;q=1, text/json;q=0.8, text/plain;q=0.5, application/x-msgpack;q=0"
            )
        headers["user-agent"] = self.sdk_configuration.user_agent

        security = (
            self.sdk_configuration.security()
            if callable(self.sdk_configuration.security)
            else self.sdk_configuration.security
        )
        client = utils.configure_security_client(
            self.sdk_configuration.client,
            security,
        )

        try:
            req = self.sdk_configuration.get_hooks().before_request(
                hook_ctx,  # type: ignore[arg-type]
                httpx.Request(
                    request.method,
                    url,
                    params=query_params,
                    data=data,
                    files=form,
                    headers=headers,
                ),
            )
            http_res = client.send(req)
        except Exception as e:  # noqa: BLE001
            _, exc = self.sdk_configuration.get_hooks().after_error(hook_ctx, None, e)  # type: ignore[arg-type]
            raise exc from e  # type: ignore[misc]

        if utils.match_status_codes(["4XX", "5XX"], http_res.status_code):
            http_res, exc = self.sdk_configuration.get_hooks().after_error(  # type: ignore[assignment]
                hook_ctx,  # type: ignore[arg-type]
                http_res,
                None,
            )
            if exc:
                raise exc
        else:
            result = self.sdk_configuration.get_hooks().after_success(
                hook_ctx,  # type: ignore[arg-type]
                http_res,
            )
            if isinstance(result, Exception):
                raise result
            http_res = result

        content_type = http_res.headers.get("Content-Type", "")

        res = response_cls(
            status_code=http_res.status_code,
            content_type=content_type,
            raw_response=http_res,
        )

        if httpx.codes.is_success(http_res.status_code):
            if utils.match_content_type(content_type, "text/plain"):
                res.content_plain = http_res.text
            elif utils.match_content_type(
                content_type,
                "application/json",
            ) or utils.match_content_type(content_type, "text/json"):
                content_cls = next(
                    field
                    for field in msgspec.structs.fields(response_cls)
                    if field.name == "content"
                ).type
                out = msgspec.json.decode(http_res.content, type=content_cls)
                res.content = out
            elif utils.match_content_type(content_type, "application/x-msgpack"):
                res.body = http_res.content
            else:
                msg = f"unknown content-type received: {content_type}"
                raise errors.CoinAPIError(
                    msg,
                    http_res.status_code,
                    http_res.text,
                    http_res,
                )
        elif codes.is_client_error(http_res.status_code) or codes.is_server_error(
            http_res.status_code,
        ):
            raise errors.CoinAPIError(
                "API error occurred",
                http_res.status_code,
                http_res.text,
                http_res,
            )

        return res
