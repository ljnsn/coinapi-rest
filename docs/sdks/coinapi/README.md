# CoinAPI SDK


## Overview

REST API: 
RESTful endpoint provides the widest range of data, based on HTTP protocol which works in Request-Reply scheme.

Implemented Standards:

 * [HTTP1.0](https://datatracker.ietf.org/doc/html/rfc1945)
 * [HTTP1.1](https://datatracker.ietf.org/doc/html/rfc2616)
 * [HTTP2.0](https://datatracker.ietf.org/doc/html/rfc7540)
 * [OpenAPI v3](https://www.openapis.org/)

> **Note:** We adhere to the OpenAPI standards for documenting our API.

## OpenAPI Specification

To access our API's OpenAPI specification, you can use the following link: [OpenAPI v3](https://raw.githubusercontent.com/coinapi/coinapi-sdk/master/data-api/coinapi-marketdata-rest.yaml)

If you need to import the OpenAPI file into software like Postman, simply copy and paste the link below:
```shell
https://raw.githubusercontent.com/coinapi/coinapi-sdk/master/data-api/coinapi-marketdata-rest.yaml
```

## Endpoints

Enviroment | Encryption | Value 
--- | --- | ---
Production | Yes | `https://rest.coinapi.io/`
Production | No | `http://rest.coinapi.io/`

:::info

For real-time market data streaming, you should use WebSockets. REST API only supports pooling, meaning you can periodically request the current market data state. In streaming, you subscribe and data or updates are delivered to you continuously.

:::

## General

If you want to learn how to authenticate to this API, you can find detailed instructions and guidance in
[authentication section](/authentication) of this documentation.

### HTTP Requests

Each HTTP request must contain the header ``Accept: application/json`` as all our responses are in JSON format.

We encourage you to use the HTTP request header ``Accept-Encoding: deflate, gzip`` for all requests. 
This will indicate to us that we can deliver compressed data to you which on your side should be decompressed transparently.

:::tip

By allowing data compression you are lowering bandwidth requirements by approximately 80%. 
This is important for requesting large amounts of data or using WebSocket Streaming API, 
as we can deliver data to you faster and more effectively.

:::

#### HTTP Success

Successful HTTP responses have the status code `200` and the body in a format according to documentation of the requested resource.

:::info

You should always check that your HTTP response status code is equal to 200, otherwise the requested was not successful.

:::

#### HTTP Errors

> Error message is returned in JSON structured like this:

```json
{
    "error": "Invalid API key"
}
```

All HTTP requests with response status codes different to `200` must be considered as failed 
and you should expect additional JSON inside the body of the response with the error message encapsulated inside it as shown in the example.
We use the following error codes:

Error Code | Meaning
---------- | -------
400 | Bad Request -- There is something wrong with your request
401 | Unauthorized -- Your API key is wrong
403 | Forbidden -- Your API key doesnt't have enough privileges to access this resource
429 | Too many requests -- You have exceeded your API key rate limits
550 | No data -- You requested specific single item that we don't have at this moment.

:::info

Good practice is to store all error messages somewhere along with request data for further manual review.

:::

### Limits

Any authenticated endpoint is providing (in HTTP response headers) information about the current state of the limits associated with API Key. In this section we will describe each limit.

#### Request limit / APIKey

```html
X-RateLimit-Limit: 1000000
X-RateLimit-Remaining: 999989
X-RateLimit-Request-Cost: 1
```

The request limit define number of maximum requests that could be executed in the 24 hours period (sliding/rollowing window - always last 24 hours from specific moment) for your subscription.

We define request as data request credits and this is not always equal to the number of API calls executed against the API. A request is deemed to be a single one if the limit query parameter on the endpoint isn’t available, isn’t used or it's stated otherwise in the API documentation. Otherwise — if the limit query parameter is available and is used — then each of the 100 data points returned in the response is counted as one request.

For example at the 2019-08-22 13:00 UTC value of the requests remaining (X-RateLimit-Remaining) will be equal to the allocated quota (X-RateLimit-Limit) decreased by the sum of the request costs (SUM(X-RateLimit-Request-Cost)) executed in the period 2019-08-21 13:00 UTC - 2019-08-22 13:00 UTC (last 24 hours).


HTTP Header | Type | Description
---------- | ------- | ---
X-RateLimit-Used | int | Provides information about the request limit that has been used within the last 24-hour period. This header indicates the amount of request capacity consumed based on the usage history. It is important to note that the header is not always appended to every request to optimize the operation of the API.
X-RateLimit-Limit | int | Is an optional feature that can be enabled via the customer portal to impose a limit on the capabilities of a specific API key. It allows you to define a threshold for the number of requests that can be made using a single API key within a 24-hour time frame.
X-RateLimit-Remaining | int | Provides information about the number of requests that can still be made within the last 24-hour period based on the usage history. This header serves as a helpful indicator of the remaining request capacity, allowing API consumers to manage their usage effectively. It is important to note that the header is not always appended to every request to optimize the operation of the API.
X-RateLimit-Request-Cost | int | The number of requests used to generate current HTTP response.
X-RateLimit-Quota-Overage | string | Provides information about whether a given API key may exceed the plan quota within a 24-hour time frame, which could result in additional charges. This header is fully defined and configured in the customer portal.
X-RateLimit-Quota-Allocated | string |  Total number of requests that can be made within a specific subscription during a 24-hour time frame. This quota allocation is determined based on the user's subscription purchase.
X-RateLimit-Quota-Remaining | string | Provides valuable information about the remaining quota within the subscription for making requests within a 24-hour time frame. This header indicates the number of requests that can still be made within the allocated quota for the current 24-hour period.

```json
GET v1/exchanges/ECB/apiKey-ED802AF4-E855-YOUR-API-KEY
Host: coinapi.io
X-RateLimit-Used: 1000
X-RateLimit-Limit: 5000
X-RateLimit-Remaining: 4000
X-RateLimit-Request-Cost: 1
X-RateLimit-Quota-Overage: ENABLED
X-RateLimit-Quota-Allocated: 10000
X-RateLimit-Quota-Remaining: 5000
```

Explanation:

- X-RateLimit-Used: 1000 (requests used in the last 24 hours)
- X-RateLimit-Limit: 5000 (total request limit within a 24-hour time frame)
- X-RateLimit-Remaining: 4000 (requests remaining within the last 24 hours)
- X-RateLimit-Request-Cost: 1 (cost or "weight" of each individual request)
- X-RateLimit-Overage: ENABLED (API key may exceed the plan quota within a 24-hour time frame)
- X-RateLimit-Quota-Allocated: 10000 (total number of requests allowed for all API keys within the subscription within a 24-hour time frame)
- X-RateLimit-Quota-Remaining: 5000 (requests remaining within the subscription's allocated quota within the last 24 hours)

#### Concurrency limit / APIKey

```html
X-ConcurrencyLimit-Limit: 10
X-ConcurrencyLimit-Remaining: 5
```

The concurrency limit defines the number of maximum concurrent API calls/requests that the API could process for your subscription at the current moment. Every API call/request increases the Concurrency limit against quota, and when it finishes, decreases it.

HTTP Header | Type | Description
---------- | ------- | ---
X-ConcurrencyLimit-Limit | int | Concurrency limit allocated for your API key.
X-ConcurrencyLimit-Remaining | int | The number of concurrent API calls/requests available to be executed in this moment for your API key.

### Output data format

By default we are using JSON output data format for all of our endpoints, you can control format of data by using `output_format` variable in query string parameters.

#### URL Parameters

Parameter | Type | Description
---------- | ------- | -------
output_format | string | Output data format *(optional, default value is `json`, possible values are `json`, `xml` or `csv`)*
csv_include_header | bool | Ignore header line in CSV output? *(optional, default value is `true`, `true` to include CSV header line, `false` otherwise)*
csv_include_quotes | bool | Encapsulate strings with quotes in CSV output? *(optional, default value is `false`, `true` to encapsulate all strings with `"`, `false` to leave them unquoted)*
csv_exclude_col | string | Comma delimited list of column names to ignore in CSV output *(optional, by default all columns are included)*
csv_set_delimiter | string | Character that will be used as column delimiter in CSV output *(optional, default value is `;`)*
csv_set_dec_mark | string | Character that will be used as decimal separator in CSV output *(optional, default value is `.`)*
csv_set_timeformat | string | Format for datetime type in CSV output or `unix` for unix timestamp *(optional, default value is `yyyy-MM-ddTHH:mm:ss.fffffffZ`)*
csv_set_newline | string | New line type *(optional, default value is `unix`, possible values `win`, `mac`, `unix`)*

### Excel / G-Sheets

There are several ways to use data from our REST API inside the Excel, Google Sheets, or similar calculation sheet application. This section will do as best as possible to keep all information up to date on how you could load the data into these applications. Feel free to contact support if we are missing an option.

#### CSV download, import:

 1. Open the data in the CSV format from the browser eg. ```https://rest.coinapi.io/v1/exchangerate/USD?apikey=YOUR_API_KEY&invert=true&output_format=csv```
 2. Save the data to the file with the .csv extension.
 3. Use the file saved and import it into the software.
 4. When configuring import, refer to the parameters like delimiter from the [Output data format](#output-data-format)

The platform-independent way described above is based on CSV but could also be used in other formats like JSON and XML as long as the software support it, but the import procedure needs to be adjusted accordingly.

#### Microsoft Excel

 * Use [PowerQuery](https://docs.microsoft.com/en-us/power-query/power-query-what-is-power-query) to load the URL directly into the CSV import without saving the file locally. 
 * Use the [=WEBSERVICE](https://support.office.com/en-us/article/webservice-function-0546a35a-ecc6-4739-aed7-c0b7ce1562c4) function to load the API response directly into the sheet, but this will not parse the data; additional processing is required.

#### Google Sheets

 * Use [=IMPORT](https://support.google.com/docs/answer/3093335?hl=en) function to load the REST API endpoint and automatically parse the CSV format data into the cells. eg. ```=IMPORTDATA("https://rest.coinapi.io/v1/exchangerate/USD?apikey=YOUR_API_KEY&invert=true&output_format=csv```

#### OpenOffice Calc

 * Select the menu Insert -> Sheet From File, 2. In the Insert dialog, put the URL eg. ```https://rest.coinapi.io/v1/exchangerate/USD?apikey=YOUR_API_KEY&invert=true&output_format=csv``` in the File Name box at the bottom. Set the drop-down list next to that to Web Page Query and click Open. The Text Import dialog opens where you can change the defaults if needed.
                    

### Available Operations

