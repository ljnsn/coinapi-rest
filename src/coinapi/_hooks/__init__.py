"""Hooks."""

from coinapi._hooks.hook_types import *
from coinapi._hooks.hooks import *
