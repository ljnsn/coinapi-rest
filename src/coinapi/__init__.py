"""CoinAPI REST SDK."""

__all__ = ("CoinAPI", "SDKConfiguration")

from coinapi.config import SDKConfiguration
from coinapi.rest import CoinAPI
