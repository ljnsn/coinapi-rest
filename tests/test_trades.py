"""Tests for trade requests."""

import pytest
from syrupy.assertion import SnapshotAssertion

from coinapi import CoinAPI
from coinapi.models import operations


@pytest.mark.vcr()
def test_get_v1_trades_symbol_id_history(
    coinapi: CoinAPI,
    snapshot: SnapshotAssertion,
) -> None:
    """Test get_v1_trades_symbol_id_history."""
    request = operations.GetV1TradesSymbolIDHistoryRequest(
        symbol_id="KRAKEN_SPOT_BTC_USD",
        time_start="2022-01-01T00:00:00",
        time_end="2022-01-02T00:00:00",
        limit=10,
    )
    response = coinapi.trades.get_v1_trades_symbol_id_history(
        request,
    )

    assert response.status_code == 200
    assert response.content is not None
    assert response.content == snapshot


@pytest.mark.vcr()
def test_get_v1_trades_symbol_id_latest(
    coinapi: CoinAPI,
    snapshot: SnapshotAssertion,
) -> None:
    """Test get_v1_trades_symbol_id_latest."""
    response = coinapi.trades.get_v1_trades_symbol_id_latest(
        symbol_id="KRAKEN_SPOT_BTC_USD",
        limit=10,
    )

    assert response.status_code == 200
    assert response.content is not None
    assert response.content == snapshot


@pytest.mark.vcr()
def test_get_v1_trades_latest(
    coinapi: CoinAPI,
    snapshot: SnapshotAssertion,
) -> None:
    """Test get_v1_trades_latest."""
    response = coinapi.trades.get_v1_trades_latest(
        filter_symbol_id="KRAKEN_SPOT_BTC_USD,KRAKEN_SPOT_ETH_USD",
        limit=10,
    )

    assert response.status_code == 200
    assert response.content is not None
    assert response.content == snapshot
