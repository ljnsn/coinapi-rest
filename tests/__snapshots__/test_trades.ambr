# serializer version: 1
# name: test_get_v1_trades_latest
  list([
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 24, 667333, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 24, 707188, tzinfo=datetime.timezone.utc), uuid='bbd51db9-d0af-4ff1-909f-fd8b2aa56272', price=67088.3, size=0.002, taker_side='SELL', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_ETH_USDT', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 23, 961974, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 23, 905651, tzinfo=datetime.timezone.utc), uuid='f2a13c82-4bc5-4264-b9cf-3a1805a84874', price=3858.6, size=0.5, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 23, 546915, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 23, 490293, tzinfo=datetime.timezone.utc), uuid='3e54b27a-d1ac-447b-9e84-15d68d7af4b7', price=67069.8, size=0.00155758, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 21, 966897, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 21, 912520, tzinfo=datetime.timezone.utc), uuid='74bcf9b8-adbb-42b4-b9c1-9e6e8e527ca2', price=67075.0, size=0.00193724, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_ETH_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 20, 72412, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 20, 15372, tzinfo=datetime.timezone.utc), uuid='aeab41b1-e4d5-4de1-b1db-c1a4ff63e5a5', price=3868.61, size=0.04846702, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 19, 898682, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 19, 841297, tzinfo=datetime.timezone.utc), uuid='1d41f10d-e595-4f67-af9c-751f68a8349d', price=67094.5, size=0.01375674, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_ETH_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 19, 467109, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 19, 406326, tzinfo=datetime.timezone.utc), uuid='bebf3041-d968-4471-aa0f-62df7042bb04', price=3868.61, size=0.00904717, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_ETH_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 17, 830938, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 17, 775935, tzinfo=datetime.timezone.utc), uuid='81740437-6145-430b-977f-d49519f21b52', price=3868.6, size=0.01, taker_side='SELL', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USDC', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 16, 446678, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 16, 390933, tzinfo=datetime.timezone.utc), uuid='ad3a764b-429c-45c0-aaf5-a3b7bbdec9e3', price=67077.76, size=0.00734322, taker_side='SELL', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 15, 378383, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 15, 319344, tzinfo=datetime.timezone.utc), uuid='e0db6288-7dc6-4f00-bc0f-fad7800abe15', price=67094.4, size=6.276e-05, taker_side='SELL', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
  ])
# ---
# name: test_get_v1_trades_symbol_id_history
  list([
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2022, 1, 1, 0, 0, 1, 151000, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2022, 1, 1, 0, 0, 1, 263989, tzinfo=datetime.timezone.utc), uuid='d00a4a46-e9f7-42d2-88a2-9d81d0282f72', price=46150.0, size=0.05, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2022, 1, 1, 0, 0, 1, 769000, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2022, 1, 1, 0, 0, 1, 884184, tzinfo=datetime.timezone.utc), uuid='7594bab3-791f-4da1-9133-c10da7204aca', price=46150.0, size=0.00268, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2022, 1, 1, 0, 0, 1, 896000, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2022, 1, 1, 0, 0, 2, 9392, tzinfo=datetime.timezone.utc), uuid='b46ee35a-9f3b-4383-8f32-ac8143ce817f', price=46150.0, size=0.00500541, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2022, 1, 1, 0, 0, 2, 914000, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2022, 1, 1, 0, 0, 3, 37176, tzinfo=datetime.timezone.utc), uuid='2f5c2a21-8604-4305-af33-df20457b4de0', price=46150.0, size=0.00376624, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2022, 1, 1, 0, 0, 3, 965000, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2022, 1, 1, 0, 0, 4, 78624, tzinfo=datetime.timezone.utc), uuid='d6994030-362e-4bf2-8ac5-d3fcee5def7f', price=46150.0, size=0.00174919, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2022, 1, 1, 0, 0, 6, 642000, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2022, 1, 1, 0, 0, 6, 755281, tzinfo=datetime.timezone.utc), uuid='a82b1f62-dbc8-49da-b379-6e40b2c0461f', price=46150.0, size=0.00015127, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2022, 1, 1, 0, 0, 8, 241000, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2022, 1, 1, 0, 0, 8, 356872, tzinfo=datetime.timezone.utc), uuid='da48580a-cf29-4760-a720-c6cb9c611010', price=46150.0, size=0.064, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2022, 1, 1, 0, 0, 8, 244000, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2022, 1, 1, 0, 0, 8, 359714, tzinfo=datetime.timezone.utc), uuid='733483c5-3b71-4fe5-861a-2d654bfa6e3f', price=46150.0, size=0.062, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2022, 1, 1, 0, 0, 8, 267000, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2022, 1, 1, 0, 0, 8, 381261, tzinfo=datetime.timezone.utc), uuid='9194397b-4d37-4e10-adf4-bde24691c995', price=46150.0, size=0.09306775, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2022, 1, 1, 0, 0, 8, 279000, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2022, 1, 1, 0, 0, 8, 398250, tzinfo=datetime.timezone.utc), uuid='2e8aaf08-1176-479f-859f-8e32e9fdba50', price=46150.0, size=0.04367, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
  ])
# ---
# name: test_get_v1_trades_symbol_id_latest
  list([
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 23, 546915, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 23, 490293, tzinfo=datetime.timezone.utc), uuid='3e54b27a-d1ac-447b-9e84-15d68d7af4b7', price=67069.8, size=0.00155758, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 21, 966897, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 21, 912520, tzinfo=datetime.timezone.utc), uuid='74bcf9b8-adbb-42b4-b9c1-9e6e8e527ca2', price=67075.0, size=0.00193724, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 19, 898682, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 19, 841297, tzinfo=datetime.timezone.utc), uuid='1d41f10d-e595-4f67-af9c-751f68a8349d', price=67094.5, size=0.01375674, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 15, 378383, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 15, 319344, tzinfo=datetime.timezone.utc), uuid='e0db6288-7dc6-4f00-bc0f-fad7800abe15', price=67094.4, size=6.276e-05, taker_side='SELL', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 4, 132549, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 4, 85993, tzinfo=datetime.timezone.utc), uuid='0642d094-32a3-478a-a07e-304ed334810d', price=67097.7, size=0.04782482, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 4, 132549, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 4, 85993, tzinfo=datetime.timezone.utc), uuid='4d703afc-3aaa-4383-a27e-be32f6374fab', price=67093.5, size=0.0155393, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 4, 132549, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 4, 85993, tzinfo=datetime.timezone.utc), uuid='8ec5d13c-89a6-474a-bdeb-2a585f93afb8', price=67084.1, size=0.00071, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 4, 132549, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 4, 85993, tzinfo=datetime.timezone.utc), uuid='8465ef74-81e8-4079-a44b-726bd67d5a07', price=67084.1, size=0.00022682, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 2, 937290, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 2, 964944, tzinfo=datetime.timezone.utc), uuid='da1bf51b-027f-4d3e-b516-5584cd373c05', price=67084.1, size=0.00077318, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
    V1Trade(symbol_id='KRAKEN_SPOT_BTC_USD', time_exchange=datetime.datetime(2024, 3, 6, 20, 46, 0, 420002, tzinfo=datetime.timezone.utc), time_coinapi=datetime.datetime(2024, 3, 6, 20, 46, 0, 366835, tzinfo=datetime.timezone.utc), uuid='a8d22d50-c072-4c0a-bc09-407128d20a80', price=67084.1, size=0.00581882, taker_side='BUY', id_trade=UNSET, id_order_maker=UNSET, id_order_taker=UNSET),
  ])
# ---
